Kernel/Devices/AudioDevice.cpp

```diff
@@ -0,0 +1,454 @@
+        return false;
+
+    size_t period_offset = card.buffer_write_offset % card.bytes_per_period;
+    size_t bytes_to_copy = card.bytes_per_period - period_offset;
+    VERIFY(bytes_to_copy <= available_space);
+    size_t request_bytes_remaining = card.current_request->buffer_size() - card.request_buffer_offset;
+    bool write_to_dma = false;
+    bool period_full = false;
+    if (bytes_to_copy > request_bytes_remaining)
+        bytes_to_copy = request_bytes_remaining;
+    else
+        period_full = true;
+
+    u8* dest_ptr;
+    if (card.dma_periods > 0 && card.buffer_write_offset - card.buffer_read_offset < card.bytes_per_period * card.dma_periods) {
+        // Either we're too close to consuming those periods or we haven't triggered playback,
+        // write directly to the dma buffer
+        dest_ptr = playback_current_dma_period(card, true) + period_offset;
+        write_to_dma = true;
+        dbgln("AudioDevice::do_transfer_request_buffer_playback writing {} bytes to dma at offset {}", bytes_to_copy, period_offset);
+    } else {
+        dest_ptr = card.buffer_region->vaddr().as_ptr();
+        dest_ptr += card.buffer_write_offset % card.bytes_all_periods;
+        dbgln("AudioDevice::do_transfer_request_buffer_playback writing {} bytes to buffer at offset {}", bytes_to_copy, period_offset);
+    }
+
+    if (!card.current_request->buffer().read(dest_ptr, bytes_to_copy)) {
+        complete_current_request(card, AsyncDeviceRequest::RequestResult::MemoryFault);
+        return false;
+    }
+    
+    card.buffer_write_offset += bytes_to_copy;
+    card.request_buffer_offset += bytes_to_copy;
+
+    if (card.state == Card::State::Prepared && period_full && (card.current.periods_trigger == 0 || card.buffer_write_offset / card.bytes_per_period == card.current.periods_trigger)) {
+        if (!period_full && card.current.periods_trigger == 0) {
+            dbgln("AudioDevice::do_transfer_request_buffer_playback triggering playback with incomplete period, will cause glitches!");
+            __builtin_memset(dest_ptr + bytes_to_copy, 0, card.bytes_per_period - bytes_to_copy); // TODO: silence samples
+            // TODO: we should probably stop playback after we run out of data and trigger an xrun instead
+        }
+        playback_dma_period_written(card, false);
+        dbgln("Triggering playback");
+        card.state = Card::State::Running;
+        trigger_playback(card);
+    } else if (period_full && write_to_dma) {
+        // Tell the driver to advance to the next dma buffer
+        playback_dma_period_written(card, card.state == Card::State::Running);
+        dbgln("in state {}", (int)card.state);
+    }
+    if (card.request_buffer_offset >= card.current_request->buffer_size()) {
+        complete_current_request(card, AsyncDeviceRequest::RequestResult::Success);
+        return false; // This doesn't indicate an error!
+    }
+    return true;
+}
+
+void AudioDevice::finished_playing_period(Card& card)
```

Comment:
Where is this supposed to be called? The NOTE comment inside mentions being called from an interrupt handler but I don't see that happening.